File: Algorithms/Implementation/Easy/Counting Valleys/Valleys.java
import java.util.Scanner;

class Valleys {
	static int CountValleys(int n, String s) {
		int valleys = 0;
		int previous_level = 0;
		int current_level = 0;

		for (int i = 0; i < n; i++) {
			previous_level = current_level;
			current_level += s.charAt(i) == 'D' ? -1 : 1;

			if ((previous_level == 0) && (current_level == -1)) {
				valleys++;
			}
		}

		return valleys;
	}

	public static void main(String[] args) {
		Scanner in = new Scanner(System.in);
		int n = in.nextInt();
		in.skip("\n");
		String s = in.nextLine();
		System.out.println(CountValleys(n, s));
	}
}

File: Algorithms/Implementation/Easy/Counting Valleys/valleys.py
def countingValleys(n, s):
    valleys = 0
    previous_level = 0
    current_level = 0
    
    for step in range(len(s)):
        previous_level = current_level
        current_level += -1 if s[step] == "D" else 1

        if previous_level == 0 and current_level == -1:
            valleys += 1
    
    return valleys


n = int(input())
s = input()
print(countingValleys(n, s))